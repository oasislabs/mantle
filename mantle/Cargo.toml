[package]
name = "mantle"
version = "0.1.0"
license = "Apache-2.0"
description = "A humane programming model for blockchain."
edition = "2018"
repository = "https://github.com/oasislabs/mantle.git"
authors = ["Oasis Labs <feedback@oasislabs.com>"]
readme = "README.md"
keywords = ["mantle"]

[dependencies]
failure = "0.1"
mantle-macros = { version = "0.1", path = "../mantle-macros" }
mantle-types = { version = "0.1", path = "../mantle-types" }
serde = { version = "1.0", features = ["derive"] }
serde_cbor = "0.9"
tiny-keccak = "1.4"

[dev-dependencies]
mantle-test = { path = "../mantle-test" }
serde = { version = "1.0", features = ["derive"] }
tiny-keccak = "1.4"

[features]
platform-alloc = []
create2 = []
