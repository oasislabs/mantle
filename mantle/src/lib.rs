#![feature(bind_by_move_pattern_guards, linkage, non_exhaustive, trait_alias)]
#![cfg_attr(target_os = "wasi", feature(wasi_ext))]

extern crate mantle_macros;

pub mod backend;
pub mod error;
pub mod exe;

pub mod reexports {
    pub use serde;
    pub use serde_cbor;
    pub use tiny_keccak;
}

pub use mantle_macros::{Event, Service};
pub use mantle_types::Address;

pub use crate::{error::Error, exe::*};

/// This macro is used to define the "main" service.
///
/// ## Example

/// ```norun
/// fn main() {
///    mantle::service!(TheMainService);
/// }
/// ```
#[macro_export]
macro_rules! service {
    ($svc:path) => {};
}

pub trait AddressExt {
    fn transfer(&self, value: u64) -> Result<(), crate::error::Error>;

    fn balance(&self) -> u64;

    fn code(&self) -> Vec<u8>;
}

impl AddressExt for Address {
    fn transfer(&self, value: u64) -> Result<(), crate::error::Error> {
        crate::backend::transact(self, value, &[]).map(|_| ())
    }

    fn balance(&self) -> u64 {
        crate::backend::balance(self).unwrap()
    }

    fn code(&self) -> Vec<u8> {
        crate::backend::code(self).unwrap()
    }
}
