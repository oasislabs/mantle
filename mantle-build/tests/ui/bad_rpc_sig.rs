use mantle::Context;

#[derive(mantle::Service)]
pub struct Counter(u32);

impl Counter {
    pub fn new(ctx: &Context) -> Self {
        Self(Default::default())
    }

    pub unsafe extern "C" fn wtf<T: std::fmt::Debug>(&self, ctx: &Context, val: T) {
        println!("val: {:?}", val);
    }
}

fn main() {
    mantle::service!(Counter);
}
