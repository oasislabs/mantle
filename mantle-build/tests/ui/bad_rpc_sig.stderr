error: RPC definition cannot have generic parameters.
  --> $DIR/bad_rpc_sig.rs:11:33
   |
11 |     pub unsafe extern "C" fn wtf<T: std::fmt::Debug>(&self, ctx: &Context, val: T) {
   |                                 ^^^^^^^^^^^^^^^^^^^^

error: RPC method cannot be unsafe.
  --> $DIR/bad_rpc_sig.rs:11:5
   |
11 | /     pub unsafe extern "C" fn wtf<T: std::fmt::Debug>(&self, ctx: &Context, val: T) {
12 | |         println!("val: {:?}", val);
13 | |     }
   | |_____^

error: RPC method cannot declare an ABI.
  --> $DIR/bad_rpc_sig.rs:11:9
   |
11 |     pub unsafe extern "C" fn wtf<T: std::fmt::Debug>(&self, ctx: &Context, val: T) {
   |         ^^^^^^^^^^^^^^^^^

error: aborting due to 3 previous errors

